torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -259,12 +263,31 @@ void ProcessGroupAgent::sync() {
 }
 
 void ProcessGroupAgent::start() {
+  start_.store(true);
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
 }
 
+void ProcessGroupAgent::localShutdown() {
+  if (!start_.load()) {
+    return;
+  }
+  LOG(INFO) << "Stopping ProcessGroupAgent.";
+  // CV to make sure we don't abort it in an invalid state.
+  {
+    std::unique_lock<std::mutex> lock(recvWorkMutex_);
+    recvWorkCV_.wait(lock, [this]() { return recvWork_; });
```

Comment:
What is this wait for?

Reply:
This ensures that `recvWork` is in a valid state (i..e. it is not uninitialized, which it would be if there was no call to `recvAnySource` in the `listenerThread`).

Reply:
It also ensures, for the thread calling stop, it gets the latests `recvWork_` that is being called `wait()`, not the previous RecvWork whose work has completed and whose `wait()` call has been unblocked.

Reply:
Wouldn't this mean that if we call shutdown without receiving any RPCs, the shutdown will just hang forever? If recvWork_ is not initialized we should just skip the `recvWork_->abort`

> It also ensures, for the user thread calling ::localShutdown(), when it gets the lock, recvWork_ is the latests RecvWork that is being called wait() on by the listenerThread_, rather than the previous RecvWork whose work has completed and whose wait() call has been unblocked.

Doesn't the lock guarantee this? Why do we need the condition_variable wait? It seems like we can get rid of this wait.